stdlib/private/SwiftPrivatePthreadExtras/PthreadBarriers.swift

```diff
@@ -73,34 +73,42 @@ public func _stdlib_pthread_barrier_init(
     return -1
   }
   barrier.pointee.mutex = UnsafeMutablePointer.allocate(capacity: 1)
-  if pthread_mutex_init(barrier.pointee.mutex!, nil) != 0 {
-    // FIXME: leaking memory.
-    return -1
-  }
   barrier.pointee.cond = UnsafeMutablePointer.allocate(capacity: 1)
-  if pthread_cond_init(barrier.pointee.cond!, nil) != 0 {
-    // FIXME: leaking memory, leaking a mutex.
+  guard _stdlib_pthread_barrier_mutex_and_cond_init(barrier) == 0 else {
+    barrier.pointee.mutex!.deinitialize()
+    barrier.pointee.mutex!.deallocate(capacity: 1)
+    barrier.pointee.cond!.deinitialize()
+    barrier.pointee.cond!.deallocate(capacity: 1)
     return -1
   }
   barrier.pointee.count = count
   return 0
 }
 
+private func _stdlib_pthread_barrier_mutex_and_cond_init(_ barrier: UnsafeMutablePointer<_stdlib_pthread_barrier_t>) -> CInt {
+  guard pthread_mutex_init(barrier.pointee.mutex!) == 0 else {
+    return -1
+  }
+  guard pthread_cond_init(barrier.pointee.cond!) == 0 else {
+    pthread_mutex_destroy(barrier.pointee.mutex!)
+    return -1
+  }
+  return 0
+}
+
 public func _stdlib_pthread_barrier_destroy(
   _ barrier: UnsafeMutablePointer<_stdlib_pthread_barrier_t>
 ) -> CInt {
-  if pthread_cond_destroy(barrier.pointee.cond!) != 0 {
-    // FIXME: leaking memory, leaking a mutex.
-    return -1
+  defer {
+    barrier.pointee.cond!.deinitialize()
+    barrier.pointee.cond!.deallocate(capacity: 1)
+    barrier.pointee.mutex!.deinitialize()
+    barrier.pointee.mutex!.deallocate(capacity: 1)
   }
-  if pthread_mutex_destroy(barrier.pointee.mutex!) != 0 {
-    // FIXME: leaking memory.
+  guard pthread_cond_destroy(barrier.pointee.cond!) == 0 &&
+    pthread_mutex_destroy(barrier.pointee.mutex!) == 0 else {
```

Comment:
Failures when releasing resources kind of weird me out. This would leak or leave the barrier in an invalid state if `pthread_cond_destroy` fails.

According to the man page, this can only fail due to API abuse (either passing an invalid parameter, or trying to destroy it while it's in use by another thread). The callers of this function either `fatalError` or signal a test failure. I'm wondering if we should just not even try to pretend that this failure can be handled gracefully, and just abort here instead.

Reply:
Good point. The caller of this function in SwiftPrivatePthreadExtras.swift does fatalError when this returns -1. Should this function simply just return Void and fatalError in here?

Reply:
I think so. If somebody wants to try to handle these errors gracefully one day, they can figure out how. No need to do that work now, only to immediately fatalError anyway.